### PR TITLE
feat: support X-Forwarded-Host header

### PR DESCRIPTION
**Pull request by MichaelMure**
Chained after https://github.com/ipfs/go-ipfs/pull/7461

This PR add support for `X-Forwarded-Host` to use the original host of the request when a proxy is used.

cc @lidel

### CODE REVIEW

**Comment by MichaelMure** [comment]
@lidel this should address all your concerns. Maybe it would need more tests ? I'll let you judge on that.

File: test/sharness/t0114-gateway-subdomains.sh [hunk 1]
```diff
@@ -664,6 +664,31 @@ test_hostname_gateway_response_should_contain \
...
+test_launch_ipfs_daemon --offline
+
+test_expect_success "request for http://fake.domain.com/ipfs/{CID} doesn't match the example.com gateway" "
+  curl -H \"Host: fake.domain.com\" -sD - \"http://127.0.0.1:$GWAY_PORT/ipfs/$CIDv1\" > response &&
+  test_should_contain \"200 OK\" response
+"
+
+test_expect_success "request for http://fake.domain.com/ipfs/{CID} with X-Forwarded-Host: example.com match the example.com gateway" "
+  curl -H \"Host: fake.domain.com\" -H \"X-Forwarded-Host: example.com\" -sD - \"http://127.0.0.1:$GWAY_PORT/ipfs/$CIDv1\" > response &&
+  test_should_contain \"Location: http://$CIDv1.ipfs.example.com/\" response
```
Comment (lidel) [inline test/sharness/t0114-gateway-subdomains.sh:689]: Do you mind also adding  `X-Forwarded-Proto: https` to the request here, and update response check to confirm it redirects to `https://`?

That way we would ensure both `X-Forwarded-Host` and `X-Forwarded-Proto` are guarded against regressions :ok_hand:

Comment (MichaelMure) [reply test/sharness/t0114-gateway-subdomains.sh:689]: Test added.

**Comment by MichaelMure** [comment]
Sharness fail is unrelated.

**Pull request merged**